src/ik_memetic.cpp

```diff
@@ -241,21 +241,118 @@ auto ik_memetic(std::vector<double> const& initial_guess,
+    // before starting to solve.
+    if (solution_fn(initial_guess)) {
+        return initial_guess;
     }
 
+    std::atomic<bool> terminate{false};
+    if (params.num_threads <= 1) {
+        // Single-threaded implementation
+        auto maybe_solution = ik_memetic_impl(initial_guess,
+                                              robot,
+                                              cost_fn,
+                                              solution_fn,
+                                              params,
+                                              terminate,
+                                              timeout,
+                                              approx_solution,
+                                              print_debug);
+        if (maybe_solution.has_value()) {
+            return maybe_solution.value().genes;
+        }
+    } else {
+        // Multi-threaded implementation
+        rsl::Queue<std::optional<Individual>> solution_queue;
+        std::vector<std::thread> ik_threads;
+        ik_threads.reserve(params.num_threads);
+
+        auto ik_thread_fn = [=, &terminate, &solution_queue]() {
+            auto soln = ik_memetic_impl(initial_guess,
+                                        robot,
+                                        cost_fn,
+                                        solution_fn,
+                                        params,
+                                        terminate,
+                                        timeout,
+                                        approx_solution,
+                                        print_debug);
+            solution_queue.push(soln);
+        };
+
+        for (size_t i = 0; i < params.num_threads; ++i) {
+            ik_threads.push_back(std::thread(ik_thread_fn));
+        }
+
+        // If enabled, stop on first solution and return.
+        size_t num_done = 0;
+        std::vector<double> best_solution;
+        auto min_cost = std::numeric_limits<double>::max();
+        if (params.stop_on_first_soln) {
+            while (solution_queue.empty() && !terminate && num_done < params.num_threads) {
+                std::this_thread::sleep_for(std::chrono::microseconds(100));
+            }
+            auto const maybe_solution = solution_queue.pop().value();
+            if (maybe_solution.has_value()) {
+                best_solution = maybe_solution->genes;
+                min_cost = maybe_solution->fitness;
+                terminate = true;
+            }
+            num_done++;
+        }
+
+        for (auto& t : ik_threads) {
+            t.join();
+        }
+
+        // Get the minimum-cost solution from all threads.
+        // Note that if approximate solutions are enabled, even if we terminate threads early, we
+        // can still compare our first solution with the approximate ones from the other threads
+        while (!solution_queue.empty()) {
+            auto const maybe_solution = solution_queue.pop().value();
```

Comment:
If there is nothing in the queue (and you don't specify a timeout) it is going the imidiatly return with a nullopt so calling value here will cause it to throw.

Here is a possible solution to that:
```suggestion
            auto const maybe_solution = solution_queue.pop(100ms);
            if (!maybe_solution) continue;
```

Reply:
Since this is happening in a `while(!solution_queue.empty())`, I think we know popping will work?

Reply:
That is a really good point, nevermind.